### ISSUE TITLE
"Node.js application" launcher doesn't expand launch variables (eg `${workspace_loc:/myProject/test.js}`)

### ISSUE THREAD

**Issue by jmini** [eclipse/wildwebdeveloper#805]
Given a Generic Eclipse project `my.project` containing a file `hello.js`:

```js
const http = require('http');

const hostname = '127.0.0.1';
const port = 3000;

const server = http.createServer((req, res) => {
  res.statusCode = 200;
  res.setHeader('Content-Type', 'text/plain');
  res.end('Hello World');
});

server.listen(port, hostname, () => {
  console.log(`Server running at http://${hostname}:${port}/`);
});
```
From [Node.js getting started guide](https://nodejs.org/en/docs/guides/getting-started-guide/)

The workspace looks like this:

<img width="913" alt="Eclipse Workspace with an open file 'hello.js'" src="https://user-images.githubusercontent.com/1222165/160100925-854921a7-ec4d-4d96-b0dd-67a57332a111.png">

---

I would like to start this using a "Node.js Application" launcher, but I am not able to configure it. 

<img width="875" alt="Node.js application launcher" src="https://user-images.githubusercontent.com/1222165/160098840-11d024e6-03e9-4e2c-98ea-89dba34a4962.png">

I always have this error `Not allowed to read file.`, no matter the value I put in `Program`.

* `node hello.js` (with the Working directory set to `${workspace_loc:/my.project/}`)
* `node ${workspace_loc:/my.project/hello.js}`
* `${workspace_loc:/my.project/hello.js}`
* `${workspaceFolder}/my.project/hello.js` 
* ...

**Comment by jmini** [comment]
As a comparison, in the same Eclipse `2022-03` I can install the old `org.eclipse.wst.jsdt.nodejs.feature` feature from the `2020-03` update site (I could not find newer version of this feature in more recent update sites).

Eclipse Installer config:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<setup.p2:P2Task
    xmi:version="2.0"
    xmlns:xmi="http://www.omg.org/XMI"
    xmlns:setup.p2="http://www.eclipse.org/oomph/setup/p2/1.0">
  <requirement
      name="org.eclipse.wst.jsdt.nodejs.feature.feature.group"
      versionRange="[1.1.0.v201908270117,1.1.0.v201908270117]"/>
  <repository
      url="https://download.eclipse.org/releases/2020-03/"/>
</setup.p2:P2Task>
```

Then it is straight forward to create the correspond launcher for the `hello.js` file:

<img width="875" alt="Node.js Application launcher from Eclipse 2020-03" src="https://user-images.githubusercontent.com/1222165/160100429-914532a7-8a97-464a-b76c-d2c5d8dec79c.png">

Launcher config as xml file:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<launchConfiguration type="org.eclipse.wst.jsdt.js.node.NodeLaunchConfigurationType">
    <booleanAttribute key="attr_add_network_console_field" value="false"/>
    <stringAttribute key="attr_app_arguments" value=""/>
    <stringAttribute key="attr_app_path" value="${workspace_loc:/my.project/hello.js}"/>
    <stringAttribute key="attr_app_project" value="my.project"/>
    <booleanAttribute key="attr_break_field" value="true"/>
    <stringAttribute key="attr_host_field" value="localhost"/>
    <stringAttribute key="attr_node_arguments" value=""/>
    <stringAttribute key="attr_port_field" value="5858"/>
    <listAttribute key="org.eclipse.debug.core.MAPPED_RESOURCE_PATHS">
        <listEntry value="/my.project"/>
    </listAttribute>
    <listAttribute key="org.eclipse.debug.core.MAPPED_RESOURCE_TYPES">
        <listEntry value="4"/>
    </listAttribute>
    <stringAttribute key="predefined_source_wrapperd_ids" value="94(org.eclipse.wst.jsdt.chromium.debug.core.model.HardcodedSourceWrapProvider$NodeJsStandardEntry)97(org.eclipse.wst.jsdt.chromium.debug.core.model.HardcodedSourceWrapProvider$NodeJsWithDefinedEntry)"/>
    <stringAttribute key="source_lookup_mode" value="EXACT_MATCH"/>
</launchConfiguration>
```

And it is working great:
<img width="992" alt="Eclipse Workspace with 'hello.js' started from Eclipse running" src="https://user-images.githubusercontent.com/1222165/160103034-652a2227-796c-4e0a-b69f-3c726f51a076.png">

**Comment by mickaelistria** [comment]
Does it work if you put an absolute path? It could be "just" variable expansion not happening for validation.

**Comment by jmini** [comment]
You are quick @mickaelistria. Thank you very much.

I started investigated.

The error message is defined here:
https://github.com/eclipse/wildwebdeveloper/blob/9c3dae600a812dc4072b95eee06dee5b8800de7b/org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/Messages.java#L28

3 usages of this constant:

```
org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/AbstractRunHTMLDebugTab.java:
  167: 			String errorMessage = Messages.RunProgramTab_error_nonReadableFile;

org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/node/RunProgramTab.java:
  71: 				String errorMessage = Messages.RunProgramTab_error_nonReadableFile;

org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/npm/NpmLaunchTab.java:
  96: 				String errorMessage = org.eclipse.wildwebdeveloper.debug.Messages.RunProgramTab_error_nonReadableFile;
```

I am not sure which one I am using in this context, but I have the feeling that all the places where this is used, does not do variable expansion… It is always: `File file = new File(programPathText.getText());`.

---

So yes you are right with absolute paths for both:

* `Program`
* `Working Directory`

Then I manage to get my server running

---

Questions:

1) When `Working Directory` is defined, shouldn't the file be relative to that folder?

2) We are sharing the `*.launch` files (we are checking them in our Git Repo so all the team members can use them). Shouln't variable expansion be available here?

**Comment by mickaelistria** [comment]
Yes, the root issue is missing variable expansion. Having relative path to working directory would be good as well.

**Comment by jmini** [comment]
During my tests I also got this (I think when the `Working Directory` was using some variables):

<img width="875" alt="Unexpected exception: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Attribute 'cwd' is not absolute ('{path}'); consider adding '${workspaceFolder}/' as a prefix to make it absolute." src="https://user-images.githubusercontent.com/1222165/160106570-dfc5fcbf-cf75-490b-94f7-d7e6f5f5bc5d.png">

```
Unexpected exception
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Attribute 'cwd' is not absolute ('{path}'); consider adding '${workspaceFolder}/' as a prefix to make it absolute.
```

But using `${workspaceFolder}` is also not supported.

**Comment by jmini** [comment]
During my tests I also got this (I think when the `Working Directory` was using some variables):

<img width="406" alt="Unexpected exception: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Attribute 'cwd' is not absolute ('{path}'); consider adding '${workspaceFolder}/' as a prefix to make it absolute." src="https://user-images.githubusercontent.com/1222165/160106956-6e2457d6-a1d4-4ae2-8412-52c4040f042b.png">

```
Unexpected exception
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Attribute 'cwd' is not absolute ('{path}'); consider adding '${workspaceFolder}/' as a prefix to make it absolute.
```

But using `${workspaceFolder}` is also not supported.

**Comment by mickaelistria** [comment]
I merged #807 that should fix that. However, the UI to suggest Variables in launch config is still missing; it's tracked in #808 .
I'm closing this one so you can leverage the fix in upcoming snapshot build. Please reopen with details if it's not resolved.
